hal/src/nRF52840/exflash_hal.cpp

```diff
@@ -0,0 +1,702 @@
+    CHECK_TRUE(ret == NRFX_SUCCESS, nrf_system_error(ret));
+
+    return 0;
+}
+
+static int gd25_secure_register_write(uint8_t security_registers_index, uint32_t security_registers_addr, const void* buf, size_t size) {
+    CHECK_TRUE(security_registers_index < GD25_SECURITY_REGISTER_COUNT, SYSTEM_ERROR_INVALID_ARGUMENT);
+    CHECK_TRUE(security_registers_addr + size <= GD25_SECURITY_REGISTER_SIZE, SYSTEM_ERROR_INVALID_ARGUMENT);
+
+    nrf_qspi_cinstr_conf_t writeSecurityRegister = PRTCL_QSPI_DEFAULT_CINSTR(HAL_QSPI_CMD_GD25_SEC_PROGRAM, NRF_QSPI_CINSTR_LEN_1B);
+
+    uint32_t address = 0;
+    uint8_t addrBuf[3] = {};
+    address = ((security_registers_index + 1) << 12) + security_registers_addr;
+    addrBuf[0] = (uint8_t)((address>>16) & 0xFF);
+    addrBuf[1] = (uint8_t)((address>>8) & 0xFF);
+    addrBuf[2] = (uint8_t)((address>>0) & 0xFF);
+
+    nrfx_err_t ret = nrfx_qspi_lfm_start(&writeSecurityRegister);
+    CHECK_TRUE(ret == NRFX_SUCCESS, nrf_system_error(ret));
+    nrf_delay_us(10); // FIXME: need a bit delay to switch between normal flash and security register
+    ret = nrfx_qspi_lfm_xfer(addrBuf, NULL, sizeof(addrBuf), false); // Send address bytes, do not finalize long frame transfer
+    CHECK_TRUE(ret == NRFX_SUCCESS, nrf_system_error(ret));
+    ret = nrfx_qspi_lfm_xfer(buf, NULL, size, true); // Send data to write, finalize long frame transfer
+    CHECK_TRUE(ret == NRFX_SUCCESS, nrf_system_error(ret));
+
+    return 0;
+}
+
+int hal_exflash_init(void) {
+    // PATCH: Initialize CS pin, external memory discharge
+    nrf_gpio_cfg_output(QSPI_FLASH_CSN_PIN);
+    nrf_gpio_pin_set(QSPI_FLASH_CSN_PIN);
+    nrf_gpio_cfg_input(QSPI_FLASH_IO1_PIN, NRF_GPIO_PIN_PULLDOWN);
+
+    FlashLock lk;
+
+    // Use single data line SPI to read ID
+    nrfx_qspi_config_t default_config = NRF_QSPI_QUICK_CFG(NRF_QSPI_WRITEOC_PP, NRF_QSPI_READOC_FASTREAD, NRF_QSPI_FREQ_32MDIV2);
+    int ret = nrfx_qspi_init(&default_config, NULL, NULL);
+    if (ret) {
+        if (ret == NRFX_ERROR_TIMEOUT) {
+            // Could have timed out because the flash chip might be in an ongoing program/erase operation.
+            // Suspend it.
+            LOG_DEBUG(TRACE, "QSPI NRFX timeout error. Suspend pgm/ers op.");
+            ret = hal_exflash_special_command(HAL_EXFLASH_SPECIAL_SECTOR_NONE, HAL_EXFLASH_COMMAND_SUSPEND_PGMERS, NULL, NULL, 0);
+        }
+        return nrf_system_error(ret);
+    }
+
+    // Determine chip type / wake up from potential sleep
+    uint8_t chip_id[3] = {};
+    nrf_qspi_cinstr_conf_t cinstr_cfg = {
+        .opcode    = HAL_QSPI_CMD_STD_READ_ID,
+        .length    = NRF_QSPI_CINSTR_LEN_4B,
+        .io2_level = true,
+        .io3_level = true,
+        .wipwait   = true,
+        .wren      = true
+    };
+    CHECK(exflash_qspi_cinstr_xfer(&cinstr_cfg, NULL, chip_id));
+
+    flash_type = hal_exflash_get_type(chip_id);
+    hal_exflash_params_t flash_params = {};
```

Comment:
We can simply define a pointer `hal_exflash_params_t*` to get  the corresponding parameters.